File: js/xssTest.js
const https = require("https");
const http  = require("http");
const { URL } = require("url");

const XSS_PAYLOADS = [
  "<script>alert('xss')</script>",
  "<img src=x onerror=alert(1)>",
  "javascript:alert(1)",
  "'><svg onload=alert(1)>",
  "<body onload=alert(1)>"
];

async function fetchURL(url, timeout = 5000) {
  return new Promise((resolve) => {
    try {
      const parsed = new URL(url);
      const lib    = parsed.protocol === "https:" ? https : http;
      const req    = lib.get(url, { timeout, rejectUnauthorized: false }, (res) => {
        let data = "";
        res.on("data", (c) => (data += c));
        res.on("end", () => resolve({ status: res.statusCode, body: data }));
      });
      req.on("error",   () => resolve(null));
      req.on("timeout", () => { req.destroy(); resolve(null); });
    } catch { resolve(null); }
  });
}

function buildTestTargets(target, siteMap) {
  const targets = new Set();
  const common  = ["q", "search", "name", "msg", "input", "query", "text"];

  for (const p of common) targets.add(`${target}?${p}=test`);

  if (siteMap?.paramUrls) {
    for (const { url } of siteMap.paramUrls) targets.add(url);
  }
  if (siteMap?.endpoints) {
    for (const ep of siteMap.endpoints) {
      for (const p of common) targets.add(`${ep}?${p}=test`);
    }
  }

  return [...targets];
}

async function testXSS(target, siteMap) {
  const result  = { name: "Cross-Site Scripting (XSS)", passed: true, issues: [] };
  const targets = buildTestTargets(target, siteMap);

  for (const baseTestUrl of targets) {
    for (const payload of XSS_PAYLOADS) {
      try {
        const parsed = new URL(baseTestUrl);
        for (const key of parsed.searchParams.keys()) {
          parsed.searchParams.set(key, payload);
        }
        const testURL = parsed.toString();
        const res     = await fetchURL(testURL);

        if (res && res.body.includes(payload)) {
          result.passed = false;
          result.issues.push({
            title:    "Reflected XSS Vulnerability",
            location: `URL: ${testURL}`,
            detail:   `Payload "${payload}" returned unescaped in HTML response.`,
            severity: "HIGH",
            risk:     "Attackers inject scripts that run in victims' browsers — stealing cookies or session tokens.",
            fixes: [
              { label: "Node.js — HTML-encode all output",
                code:  "const he = require('he');\nres.send('<p>' + he.encode(userInput) + '</p>');" },
              { label: "React — never use dangerouslySetInnerHTML",
                code:  "// ✅ Safe — React escapes by default\n<p>{userInput}</p>" }
            ]
          });
          break;
        }
      } catch { }
    }
  }

  return result;
}

module.exports = testXSS;

File: C#/XssTest.cs
namespace SecurityTester.Tests;

public class XssTest
{
    private static readonly string[] Payloads =
    {
        "<script>alert('xss')</script>",
        "<img src=x onerror=alert(1)>",
        "javascript:alert(1)",
        "'><svg onload=alert(1)>",
        "<body onload=alert(1)>"
    };

    private static readonly string[] CommonParams =
        { "q", "search", "name", "msg", "input", "query", "text", "s" };

    private static List<string> BuildTargets(string target, SiteMap? siteMap)
    {
        var targets = new HashSet<string>(StringComparer.OrdinalIgnoreCase);

        foreach (var p in CommonParams)
            targets.Add($"{target}?{p}=test");

        if (siteMap == null) return targets.ToList();

        foreach (var page in siteMap.ParamUrls)
            targets.Add(page.Url);

        foreach (var ep in siteMap.Endpoints)
            foreach (var p in CommonParams)
                targets.Add($"{ep}?{p}=test");

        return targets.ToList();
    }

    public static async Task<TestResult> RunAsync(string target, HttpClient client, SiteMap? siteMap = null)
    {
        var result  = new TestResult { Name = "Cross-Site Scripting (XSS)" };
        var targets = BuildTargets(target.TrimEnd('/'), siteMap);

        foreach (var baseUrl in targets)
        {
            foreach (var payload in Payloads)
            {
                try
                {
                    var uri = new Uri(baseUrl);
                    var qs  = System.Web.HttpUtility.ParseQueryString(uri.Query);
                    foreach (var key in qs.AllKeys) qs[key] = payload;

                    var testUrl = $"{uri.GetLeftPart(UriPartial.Path)}?{qs}";
                    var res     = await client.GetAsync(testUrl);
                    var body    = await res.Content.ReadAsStringAsync();

                    if (body.Contains(payload))
                    {
                        result.Passed = false;
                        result.Issues.Add(new SecurityIssue
                        {
                            Title    = "Reflected XSS Vulnerability",
                            Location = $"URL: {testUrl}",
                            Detail   = $"Payload \"{payload}\" returned unescaped in response.",
                            Severity = "HIGH",
                            Risk     = "Attackers inject scripts into victims' browsers — stealing sessions or cookies.",
                            Fixes    = new()
                            {
                                new() { Label = "C# — HTML-encode all output",
                                        Code  = "using System.Text.Encodings.Web;\n\n// ❌ VULNERABLE\nResponse.Write(\"<p>\" + userInput + \"</p>\");\n\n// ✅ SAFE\nResponse.Write(\"<p>\" + HtmlEncoder.Default.Encode(userInput) + \"</p>\");" },
                                new() { Label = "Razor — never use Html.Raw with user input",
                                        Code  = "// ❌ DANGEROUS\n@Html.Raw(userInput)\n\n// ✅ SAFE — Razor escapes automatically\n@userInput" }
                            }
                        });
                        goto nextTarget;
                    }
                }
                catch { }
            }
            nextTarget:;
        }

        return result;
    }
}
